Constrained Beam Search [without disjunctive decoding]

cwkeam opened the pull request:
# ~~Disjunctive~~ Positive Constraint Decoding 

@patrickvonplaten @Narsil 

Fixes #14081, 

But without the "disjunctive" component. I think it's a good idea to first  integrate this `Constraint` framework before adding more advanced techniques like Disjunctive Decoding, which can arguably be added with ease afterwards.

Currently this PR is able to force the generation of tokens and/or an ordered sequence of tokens in the beam search with `constrained_beam_search` in `GenerationMixin.generate`. It finds the best possible, highest probable list of beam outputs in fulfilling a list of `Constraint` objects, yielding results more sensible than just blindly forcing the tokens in the output sequence.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. #14081 
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

Tagging @patrickvonplaten because my main back-and-forth was with you on this PR.

## Example Use

Here is an example of how one could use this functionality:

```python
from transformers import GPT2Tokenizer, GPT2LMHeadModel
from transformers.generation_beam_constraints import PhrasalConstraint

device = "cuda"

model = GPT2LMHeadModel.from_pretrained("gpt2").to(device)
tokenizer = GPT2Tokenizer.from_pretrained("gpt2")

force_text = " big monsters"
force_text_2 = " crazy"
force_tokens = tokenizer.encode(force_text, return_tensors="pt").to(device)[0]
force_tokens_2 = tokenizer.encode(force_text_2, return_tensors="pt").to(device)[0]

constraints = [
    PhrasalConstraint(force_tokens),
    PhrasalConstraint(force_tokens_2)
]

input_text = ["The baby is crying because"] 

model_inputs = tokenizer(input_text, return_tensors="pt")

for key, value in model_inputs.items():
    model_inputs[key] = value.to(device)

k = model.generate(
    **model_inputs,
    constraints=constraints,
    num_beams=7,
    num_return_sequences=7,
    no_repeat_ngram_size=2
)

for out in k:
    print(tokenizer.decode(out))
```
For some example outputs:
```
The baby is crying because she's been told crazy big monsters are going to come and kill her.
The baby is crying because she's been told crazy big monsters are coming for her.
The baby is crying because she's been told crazy big monsters are going to come after her.
```

## The Constraint Framework

Users can define their own constraints by inheriting the `Constraint` abstract class and this framework is ensured to work as desired, because the `Constraint` class is quite strictly defined. If an implementation passes the `self.test()` function of this interface then it necessarily works as desired. An incorrect implementation will lead to an error. But probably most users would be fine just using the `TokenConstraint` and `PhrasalContraint` classes.

## Testing

My approach to testing this new functionality was that it passes every test that was related to the existing `beam_search` function, with the **additional testing that requires that the generated output indeed fulfills the constraint.**

One such example:
```python
# tests/test_generation_utils.py 

def test_constrained_beam_search_generate(self):
    for model_class in self.all_generative_model_classes:
        ....
        # check `generate()` and `constrained_beam_search()` are equal for `num_return_sequences`
        # Sample constraints
        force_tokens = torch.randint(min_id, max_id, (1, 2)).type(torch.LongTensor)[0]
        constraints = [
            PhrasalConstraint(force_tokens),
        ]
        
        num_return_sequences = 2
        max_length = 10
        
        beam_kwargs, beam_scorer = self._get_constrained_beam_scorer_and_kwargs(
            input_ids.shape[0], max_length, constraints, num_return_sequences=num_return_sequences
        )
        
        output_generate, output_beam_search = self._constrained_beam_search_generate(
            model=model,
            input_ids=input_ids,
            attention_mask=attention_mask,
            max_length=max_length,
            constrained_beam_scorer=beam_scorer,
            constraints=constraints,
            beam_kwargs=beam_kwargs,
            logits_processor=logits_processor,
            logits_process_kwargs=logits_process_kwargs,
        )
        self.assertListEqual(output_generate.tolist(), output_beam_search.tolist())
        
        for generation_output in output_generate:
            self._check_sequence_inside_sequence(force_tokens, generation_output)
...
```

cwkeam committed:
added classes to get started with constrained beam search

cwkeam committed:
in progress, think i can directly force tokens now but not yet with the round robin

cwkeam committed:
think now i have total control, now need to code the bank selection

cwkeam committed:
technically works as desired, need to optimize and fix design choices leading to undersirable outputs

cwkeam committed:
Merge branch 'huggingface:master' into master

cwkeam committed:
complete PR #1 without disjunctive decoding

cwkeam committed:
Merge branch 'master' of https://github.com/cwkeam/transformers

cwkeam committed:
removed incorrect tests

cwkeam committed:
Delete k.txt

cwkeam committed:
Delete test.py

cwkeam committed:
Delete test.sh

cwkeam committed:
revert changes to test scripts

cwkeam committed:
Merge branch 'master' of https://github.com/cwkeam/transformers

cwkeam committed:
Merge branch 'master' into master

cwkeam committed:
genutils

cwkeam committed:
Merge branch 'master' of https://github.com/cwkeam/transformers

cwkeam committed:
full implementation with testing, no disjunctive yet

cwkeam committed:
shifted docs

cwkeam committed:
Merge branch 'huggingface:master' into constrained_beam_search

cwkeam committed:
passing all tests realistically ran locally

HuggingFaceDocBuilder commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_15416). All of your documentation changes will be reflected on that endpoint.

cwkeam committed:
removing accidentally included print statements

cwkeam committed:
fixed source of error in initial PR test

cwkeam committed:
fixing the get_device() vs device trap

cwkeam committed:
fixed documentation docstrings about constrained_beam_search

patrickvonplaten commented:
Very cool taking a look now

patrickvonplaten commented on src/transformers/generation_utils.py:
```diff
@@ -1326,6 +1333,41 @@ def generate(
                 **model_kwargs,
             )
 
+        elif is_constraint_gen_mode:
```
If I understand correctly, this makes only sense if:
- `num_beams` > 1 
- `do_sample = False`

no? 

Could we maybe raise a `ValueError` if any of the above conditions is not fullfilled? This would also greatly help us reduce maintenance in a first step since we're limiting the amount of possible use cases. We could always later add contrained generation for sampling or non-beam search. What do you think?

cwkeam replied:
Agreed. I'll adjust accordingly in a following commit.

patrickvonplaten commented on src/transformers/generation_utils.py:
```diff
@@ -2787,6 +2829,315 @@ def group_beam_search(
+        ...     )
+        ... }
+
+        >>> constraints = [PhrasalConstraint(tokenizer.encode("required phrase")[0])]
```
It's not super obvious to me what this contraint would do actually. Could we maybe use a "real" contraint here?

cwkeam replied:
What do you mean by a "real" constraint as opposed to what is in the above example? 

I do share the general concern about the lack of clarity on what specific effect these "constraints" have on the generation output. As per the other comment below about showing the actual output. I'll rework the documentation again in a following commit.

patrickvonplaten commented on src/transformers/generation_utils.py:
```diff
@@ -2787,6 +2829,315 @@ def group_beam_search(
+        ...     input_ids, beam_scorer, constraints=constraints, logits_processor=logits_processor, **model_kwargs
+        ... )
+
+        >>> print("Generated:", tokenizer.batch_decode(outputs, skip_special_tokens=True))
```
```suggestion
        >>> print("Generated:", tokenizer.batch_decode(outputs, skip_special_tokens=True))
        # => <the-actual-output>
```

Could we return the actual output here? Think this would be very nice for users so that they see directly when it makes sense to use constrained beam search

patrickvonplaten replied:
Nice!

cwkeam committed:
fixed tests having failing for Speech2TextModel's floating point inputs

patrickvonplaten committed:
fix cuda long tensor

patrickvonplaten committed:
Merge branch 'constrained_beam_search' of https://github.com/cwkeam/transformers into constrained_beam_search

patrickvonplaten commented:
Hey @cwkeam,

This already looks like a super clean and impactful addition to the `transformers` library! Great job. I really like how you implement this new generation method in a extendable way without affecting existing generate methods. I'm sure we can merge it soon. 

At the moment, it seems like your example above doesn't work. When I try to run:

```python
from transformers import GPT2Tokenizer, GPT2LMHeadModel
from transformers.generation_beam_constraints import PhrasalConstraint

device = "cuda"

model = GPT2LMHeadModel.from_pretrained("gpt2").to(device)
tokenizer = GPT2Tokenizer.from_pretrained("gpt2")

force_text = " big monsters"
force_text_2 = " crazy"
force_tokens = tokenizer.encode(force_text, return_tensors="pt").to(device)[0]
force_tokens_2 = tokenizer.encode(force_text_2, return_tensors="pt").to(device)[0]

constraints = [
    PhrasalConstraint(force_tokens),
    PhrasalConstraint(force_tokens_2)
]

input_text = ["The baby is crying because"] 

model_inputs = tokenizer(input_text, return_tensors="pt")

for key, value in model_inputs.items():
    model_inputs[key] = value.to(device)

k = model.generate(
    **model_inputs,
    constraints=constraints,
    num_beams=7,
    num_return_sequences=7,
    no_repeat_ngram_size=2
)

for out in k:
    print(tokenizer.decode(out))
```

on my computer, I'm getting the following error:

```
ValueError: `token_ids` has to be list of positive integers or a `torch.LongTensor` but is tensor([7165], device='cuda:0')For single token constraints, refer to `TokenConstraint`.
```

I think the example is really nice and exactly what we are looking for. So maybe you could also add it as a `@slow` test under `tests/test_generation_utils.py` once it works :-)

patrickvonplaten commented:
Let me know if you need further help!

cwkeam committed:
added examples and testing for them and founx & fixed a bug in beam_search and constrained_beam_search

cwkeam committed:
merge fix

cwkeam commented:
Added 1. clear examples (**and tests for them**, 2. more specific input parameter tests, and 3. **bug fixes in other beam search code**.

## 1. Examples (text completion & translation)

The latter example really emphasizes the usefulness of this feature.

Fixed up the first example as requested previously, showing its ability that it can

- [x] handle multiple constraints
- [x] handle constraints that have more than one words (sequence of tokens)

```python
def test_constrained_beam_search(self):
    model = GPT2LMHeadModel.from_pretrained("gpt2").to(torch_device)
    tokenizer = GPT2Tokenizer.from_pretrained("gpt2")

    force_tokens = tokenizer.encode(" scared", return_tensors="pt").to(torch_device)[0]
    force_tokens_2 = tokenizer.encode(" big weapons", return_tensors="pt").to(torch_device)[0]

    constraints = [
        PhrasalConstraint(force_tokens),
        PhrasalConstraint(force_tokens_2),
    ]

    starting_text = ["The soldiers were not prepared and"]

    input_ids = tokenizer(starting_text, return_tensors="pt").input_ids.to(torch_device)

    outputs = model.generate(
        input_ids,
        constraints=constraints,
        num_beams=10,
        num_return_sequences=1,
        no_repeat_ngram_size=1,
        max_length=30,
        remove_invalid_values=True,
    )

    generated_text = tokenizer.batch_decode(outputs, skip_special_tokens=True)

    self.assertListEqual(
        generated_text,
        [
            "The soldiers were not prepared and didn't know how big the big weapons would be, so they scared them off. They had no idea what to do",
        ],
    )

```

and here's a translation example:

```python
def test_constrained_beam_search_example_integration(self):
    tokenizer = AutoTokenizer.from_pretrained("t5-base")
    model = AutoModelForSeq2SeqLM.from_pretrained("t5-base")

    encoder_input_str = "translate English to German: How old are you?"
    encoder_input_ids = tokenizer(encoder_input_str, return_tensors="pt").input_ids

    # lets run beam search using 5 beams
    num_beams = 5
    # define decoder start token ids
    input_ids = torch.ones((num_beams, 1), device=model.device, dtype=torch.long)
    input_ids = input_ids * model.config.decoder_start_token_id

    # add encoder_outputs to model keyword arguments
    model_kwargs = {
        "encoder_outputs": model.get_encoder()(
            encoder_input_ids.repeat_interleave(num_beams, dim=0), return_dict=True
        )
    }

    constraint_str = "sind"
    constraint_token_ids = tokenizer.encode(constraint_str)[:-1]  # remove eos token
    constraints = [PhrasalConstraint(token_ids=constraint_token_ids)]

    # instantiate beam scorer
    beam_scorer = ConstrainedBeamSearchScorer(
        batch_size=1, num_beams=num_beams, device=model.device, constraints=constraints
    )

    # instantiate logits processors
    logits_processor = LogitsProcessorList(
        [
            MinLengthLogitsProcessor(5, eos_token_id=model.config.eos_token_id),
        ]
    )

    outputs = model.constrained_beam_search(
        input_ids, beam_scorer, constraints=constraints, logits_processor=logits_processor, **model_kwargs
    )
    outputs = tokenizer.batch_decode(outputs, skip_special_tokens=True)

    self.assertListEqual(outputs, ["Wie alter sind Sie?"])

```

This is a really useful example, because the following is the test for **ordinary beam search**

```python
def test_beam_search_example_integration(self):
    # ... the exact same code as above minus constraints
    outputs = model.beam_search(input_ids, beam_scorer, logits_processor=logits_processor, **model_kwargs)
    outputs = tokenizer.batch_decode(outputs, skip_special_tokens=True)

    self.assertListEqual(outputs, ["Wie alt bist du?"])

```

I've never studied German before, but upon looking it up, it seemed like there were two equally realistic ways to translate "How old are you" in German. While the ordinary beam search output may be fine for most uses, maybe the user specifically wants the other way of translating it. In this case, they can do as I have and force the generation to include the token that changes the direction of the translation as desired.

## 2. Input Parameters tests

Added @patrickvonplaten 's recommendation as well as dealing with `num_groups`:

```python
if num_beams <= 1:
    raise ValueError("`num_beams` needs to be greater than 1 for constrained genertation.")

if do_sample:
    raise ValueError("`do_sample` needs to be false for constrained generation.")

if num_beam_groups is not None and num_beam_groups > 1:
    raise ValueError("`num_beam_groups` not supported yet for constrained generation.")
```

I'm not sure whether the last part is a good idea and whether I should just make the implementation work for `num_beam_groups>1` but the fact is that it can't handle that case yet (will throw an error).

## 3. Bug Fixes in **beam_search** with `max_length`

As I was writing the test code for the few examples above, I realized that the example in the huggingface documentation for **ordinary beam search** was actually failing, because of the **recent change with `max_length`**.

The above testing code `test_beam_search_example_integration(self)` is taken directly from the huggingface documentation for `GenerationMixin.beam_search` and it fails, because `max_length` can be set to None unless otherwise specified, and it throws an error saying that you can't compare a `Tensor` to a `NoneType`.

[From `BeamSearchScorer.finalize()`](https://github.com/huggingface/transformers/blob/master/src/transformers/generation_beam_search.py#L333),

```python
sent_max_len = min(sent_lengths.max().item() + 1, max_length) # ERROR 1
decoded: torch.LongTensor = input_ids.new(batch_size * self.num_beam_hyps_to_keep, sent_max_len)
if sent_lengths.min().item() != sent_lengths.max().item():
    assert pad_token_id is not None, "`pad_token_id` has to be defined"
    decoded.fill_(pad_token_id)

for i, hypo in enumerate(best):
    decoded[i, : sent_lengths[i]] = hypo
    if sent_lengths[i] < max_length: # ERROR 2
        decoded[i, sent_lengths[i]] = eos_token_id
```

So I changed that to:

```python
sent_lengths_max = sent_lengths.max().item() + 1
sent_max_len = min(sent_lengths_max, max_length) if max_length is not None else sent_lengths_max
decoded: torch.LongTensor = input_ids.new(batch_size * self.num_beam_hyps_to_keep, sent_max_len)
# shorter batches are padded if needed
if sent_lengths.min().item() != sent_lengths.max().item():
    assert pad_token_id is not None, "`pad_token_id` has to be defined"
    decoded.fill_(pad_token_id)

# fill with hypotheses and eos_token_id if the latter fits in
for i, hypo in enumerate(best):
    decoded[i, : sent_lengths[i]] = hypo
    if sent_lengths[i] < sent_max_len:
        decoded[i, sent_lengths[i]] = eos_token_id
```

cwkeam committed:
deleted accidentally added test halting code with assert False

cwkeam committed:
code reformat

cwkeam commented:
Hey @patrickvonplaten 

I also wanted to ask your opinions on making this feature integrate more easily to the `model.generate()` function. The same way how `bad_word_ids` is converted into a `LogitsProcessor` in the internal logic, maybe there's a way to add a similarly simple API like `force_token_ids` that internally converts it into `Constraint` objects in the backend.

I'm thinking of something like:

```python
starting_text = ["The soldiers were not prepared and"]
force_tokens = ["scared", "big weapons"]

input_ids = tokenizer(starting_text, return_tensors="pt").input_ids.to(torch_device)
force_token_ids = tokenizer(force_tokens).input_ids

outputs = model.generate(
    input_ids,
    force_token_ids=force_token_ids, # [[28349], [23981, 3721]] something that looks like this (made the numbers up)
   ...
)
```

But the problem arises when we start talking about **Disjunctive Positive Constraint Decoding** which was the main subject of the original #14081 issue. Although the disjunctive constraint generation implementation isn't even made yet, I think the best approach is something like this:

```python
from transformers.generation_beam_constraints import DisjunctiveConstraint

variations_1 = ["rain", "raining", "rained"]
variations_2 = ["smile", "smiling", "smiled"]

force_rain = tokenizer.encode(variations, return_tensors="pt")
force_smile = tokenizer.encode(variatiosn_2, return_tensors="pt")

constraints = [
    DisjunctiveConstraint(force_rain),
    DisjunctiveConstraint(force_smile)
]

outputs = model.generate(
    input_ids,
    constraints=constraints,
   ...
)
```

And this feature would be something that's super awkward to make into the `force_token_ids` API.

I think one way to approaching this is realistinc that a `PhrasalConstraint` is basically a subset of a `DisjunctiveConstraint`:

```python
variations_1 = ["rain"]
variations_2 = ["smile"]

force_rain = tokenizer.encode(variations, return_tensors="pt")
force_smile = tokenizer.encode(variatiosn_2, return_tensors="pt")

constraints = [
    DisjunctiveConstraint(force_rain),
    DisjunctiveConstraint(force_smile)
]
```

So maybe the API above should be changed to:

```python
starting_text = ["The soldiers were not prepared and"]
force_tokens = ["scared", "big weapons"]

input_ids = tokenizer(starting_text, return_tensors="pt").input_ids.to(torch_device)
force_token_ids = tokenizer(force_tokens).input_ids

outputs = model.generate(
    input_ids,
    force_token_ids=force_token_ids, # [
        [ [28349], [456849] ], # DisjunctiveConstraint, choose one between two
        [ [23981, 3721] ] # ordinary PhrasalConstraint for a sequence
     ]  
   ...
)
```

Then the input type for `force_token_ids` would be `List[List[List[int]]]` that sounds quite messy.

Let me know your thoughts on this!

patrickvonplaten commented on tests/test_generation_utils.py:
```diff
@@ -2047,3 +2309,120 @@ def test_transition_scores_group_beam_search_encoder_decoder(self):
 
         self.assertTrue(torch.allclose(transition_scores_sum, outputs.sequences_scores, atol=1e-3))
+
+    def test_constrained_beam_search(self):
```
```suggestion
    @slow
    def test_constrained_beam_search(self):
```

patrickvonplaten commented on tests/test_generation_utils.py:
```diff
@@ -2047,3 +2309,120 @@ def test_transition_scores_group_beam_search_encoder_decoder(self):
+
+        self.assertListEqual(outputs, ["Wie alt bist du?"])
+
+    def test_constrained_beam_search_example_integration(self):
```
```suggestion
    @slow
    def test_constrained_beam_search_example_integration(self):
```

patrickvonplaten reviewed (commented):
Hey @cwkeam,

The PR looks to be in a great shape. I can reproduce the examples that you added - very nice. I agree that we should make it as easy as possible to use the functionality by adding a keyword argument to generate, but I think we can do this in a follow up PR to make reviewing easier and to make sure we don't have to change any naming in the future.

Would it be ok for you to for now to **not** add a keyword argument and to do this in a follow up PR?

Test tests you added look great so I think now we just need to get the Circle CI green and this should be good for a final review :-)

cwkeam committed:
Update tests/test_generation_utils.py

Co-authored-by: Patrick von Platen <patrick.v.platen@gmail.com>

cwkeam committed:
Update tests/test_generation_utils.py

Co-authored-by: Patrick von Platen <patrick.v.platen@gmail.com>

cwkeam committed:
Update tests/test_generation_utils.py

Co-authored-by: Patrick von Platen <patrick.v.platen@gmail.com>

cwkeam committed:
Update tests/test_generation_utils.py

Co-authored-by: Patrick von Platen <patrick.v.platen@gmail.com>

cwkeam committed:
Merge branch 'huggingface:master' into constrained_beam_search

cwkeam commented:
Hey @patrickvonplaten,

That's a good point. I think that the `DisjunctiveConstraint` feature should also come as a separate PR for maintanence purposes as you said. I'll open a new PR soon.

The checks seem to be all passing now! Will you do a final review?

Thanks!

patrickvonplaten commented on src/transformers/generation_beam_search.py:
```diff
@@ -330,7 +367,430 @@ def finalize(
+        decoded: torch.LongTensor = input_ids.new(batch_size * self.num_beam_hyps_to_keep, sent_max_len)
+        # shorter batches are padded if needed
+        if sent_lengths.min().item() != sent_lengths.max().item():
+            assert pad_token_id is not None, "`pad_token_id` has to be defined"
```
Could we please transform this into:
```suggestion
            if pad_token_id is None:
                raise ValueError("`pad_token_id` has to be defined")
```

patrickvonplaten replied:
This code is very sensitive and should not be changed at all. Is GitHub just showing a very bad diff or was some code changed here?

patrickvonplaten commented on src/transformers/generation_beam_search.py:
```diff
@@ -330,7 +367,430 @@ def finalize(
+        decoded: torch.LongTensor = input_ids.new(batch_size * self.num_beam_hyps_to_keep, sent_max_len)
+        # shorter batches are padded if needed
+        if sent_lengths.min().item() != sent_lengths.max().item():
+            assert pad_token_id is not None, "`pad_token_id` has to be defined"
```
Ok actually this looks good. Just seems like GitHub messed up the diff, no?

patrickvonplaten replied:
Ok no actually there are small changes to the original `BeamScorer.finalize(...)` function. E.g.:
` sent_max_len = min(sent_lengths.max().item() + 1, max_length)` is replaced by `sent_max_len = min(sent_lengths_max, max_length) if max_length is not None else ...` and `if sent_lengths[i] < max_length:` by `if sent_lengths[i] < sent_max_len`. 

@cwkeam - could you please revert those changes? This looks very much backward breaking. Beam search is very complex and small changes like those might lead to different results even if our tests all run correctly. Could you please make sure that nothing is changed to the existing `BeamScorer` class?

@cwkeam - could you revert this changes

cwkeam replied:
@patrickvonplaten Thinking back, I do agree that this should not have been added to this PR. However, this is actually a solution to a bug I found in the existing implementation as explained here: https://github.com/huggingface/transformers/pull/15416#issuecomment-1028550095

Looking back, I should've submitted this with a separate PR. I'll revert these changes for now.

cwkeam replied:
Opened a PR to track this issue #15555 .

patrickvonplaten commented on tests/test_generation_utils.py:
```diff
@@ -2047,3 +2309,124 @@ def test_transition_scores_group_beam_search_encoder_decoder(self):
         self.assertTrue(torch.allclose(transition_scores_sum, outputs.sequences_scores, atol=1e-3))
+
+    @slow
+    @slow
```
```suggestion
```

patrickvonplaten reviewed (approved):
Looks good to me. IMO we can merge this. 

@Narsil , @patil-suraj @sgugger @LysandreJik - could you take a very high level look as well?

patrickvonplaten committed:
Update tests/test_generation_utils.py

patrickvonplaten commented:
Incredible job at implementing this @cwkeam - it's very clean and enables some really cool use cases. 
If you'd like we could now the Conjunctive decoding in a follow-up PR then and make it a bit easier for the user to enable it via some keywords for `generate()` and then post about this method - I'm sure the community will be very excited about this addition. If you'd like to promote the integration already at this point, we could also do it right away and delay the conjuntive decoding PR a bit - as you wish :-)

patrickvonplaten reviewed (changes requested):
See https://github.com/huggingface/transformers/pull/15416/files#r800972360 - after than we can merge I think

patrickvonplaten commented:
All slow tests run correctly and the fast tests are not slowed down by this PR. One final change @cwkeam and I think then we are good to go

sgugger commented on src/transformers/generation_beam_constraints.py:
```diff
@@ -0,0 +1,372 @@
+            if counter == 1:
+                self.reset()
+            advance = self.advance()
+            assert self.does_advance(advance)
```
We don't add new `assert` in the code base, especially not without an error message. Please put a test an raise the appropriate exception here.

sgugger commented on src/transformers/generation_beam_constraints.py:
```diff
@@ -0,0 +1,372 @@
+            if counter > 10000:
+                raise Exception("update() does not fulfill the constraint.")
+
+        assert self.remaining() == 0
```
Same here.

sgugger commented on src/transformers/generation_beam_constraints.py:
```diff
@@ -0,0 +1,372 @@
+            token_id(`int`):
+                The id of a newly generated token in the beam search.
+        returns:
+            stepped(`boolean`):
+                Whether this constraint has become one step closer to being fulfuilled.
+            completed(`boolean`):
+                Whether this constraint has been completely fulfilled by this token being generated.
+            reset (`boolean`):
```
The type is named `bool` in Python.

sgugger commented on src/transformers/generation_beam_constraints.py:
```diff
@@ -0,0 +1,372 @@
+        Creates a new instance of this constraint.
+
+        Args:
+            stateful(`boolean`): Whether to not only copy the constraint for new instance, but also its state.
```
Same here
```suggestion
            stateful(`bool`): Whether to not only copy the constraint for new instance, but also its state.
```

sgugger commented on src/transformers/generation_beam_constraints.py:
```diff
@@ -0,0 +1,372 @@
+
+
+class PhrasalConstraint(Constraint):
+    @add_start_docstrings(PHRASAL_CONSTRAINT_DOCSTRING)
```
As said above, just put the docstring here.

sgugger commented on src/transformers/generation_beam_constraints.py:
```diff
@@ -0,0 +1,372 @@
+
+# For beam scorers to track its progress through a list of constraints.
+class ConstraintListState:
+    def __init__(self, constraints: List[Constraint]):
```
This deserves a docstring instead of the small comment above.

sgugger commented on src/transformers/generation_beam_constraints.py:
```diff
@@ -0,0 +1,372 @@
+            return complete, stepped
+
+        if self.inprogress_constraint is not None:
+            """
+            In the middle of fulfilling a constraint. If the `token_id` *does* makes an incremental progress to current
+            job, simply update the state
+            """
```
This should be a comment.

sgugger commented on src/transformers/generation_beam_constraints.py:
```diff
@@ -0,0 +1,372 @@
+            """
+            stepped, complete, reset = self.inprogress_constraint.update(token_id)
+            if reset:
+                """
+                1. If the next token breaks the progress, then we must restart.
+                    e.g. constraint = "I love pies" and sequence so far is "I love" but `token_id` == "books".
+
+                    But that doesn't mean we self.init_state(), since we only reset the state for this particular
+                    constraint, not the full list of constraints.
+                """
```
Same here

sgugger commented on src/transformers/generation_beam_constraints.py:
```diff
@@ -0,0 +1,372 @@
+                self.inprogress_constraint = None
+
+            if complete:
+                """
+                2. If the next token completes the constraint, move it to completed list, set
+                    inprogress to None. If there are no pending constraints either, then this full list of constraints
+                    is complete.
+                """
```
Same there.

sgugger commented on src/transformers/generation_beam_constraints.py:
```diff
@@ -0,0 +1,372 @@
+                    self.completed = True
+
+        else:
+            """
+            Not in the middle of fulfilling a constraint. So does this `token_id` helps us step towards any of our list
+            of constraints?
+            """
```
Same here.

sgugger commented on src/transformers/generation_beam_constraints.py:
```diff
@@ -0,0 +1,372 @@
+                        self.inprogress_constraint = pending_constraint
+
+                    if complete or stepped:
+                        """
+                        If we made any progress at all, then it's at least not a "pending constraint".
+                        """
```
Same there

sgugger commented on src/transformers/generation_beam_constraints.py:
```diff
@@ -0,0 +1,372 @@
+                        )
+
+                        if len(self.pending_constraints) == 0 and self.inprogress_constraint is None:
+                            """
+                            If there's no longer any pending after this and no inprogress either, then we must be
+                            complete.
+                            """
```
Same there.

sgugger commented on src/transformers/generation_beam_search.py:
```diff
@@ -83,6 +85,41 @@
 """
 
 
+CONSTRAINT_PROCESS_INPUTS_DOCSTRING = r"""
```
Same as before, this is only used once, so it should directly go on the method it documents.

sgugger commented on src/transformers/generation_beam_search.py:
```diff
@@ -330,7 +367,430 @@ def finalize(
+        sent_beam_indices: torch.LongTensor,
+        push_progress: bool = False,
+    ):
+        # from transformers import GPT2Tokenizer
+        # tokenizer = GPT2Tokenizer.from_pretrained("gpt2")
```
To clean up?

sgugger commented on src/transformers/generation_beam_search.py:
```diff
@@ -330,7 +367,430 @@ def finalize(
+    ):
+        # from transformers import GPT2Tokenizer
+        # tokenizer = GPT2Tokenizer.from_pretrained("gpt2")
+        """
+        sent_beam_tokens are the next {num_beams} number of tokens that are under consideration for this beam
+        (candidate next tokens)
+
+        1. Adding "advance_tokens"
+            using ConstraintStateList.advance(), we propose new tokens to be added into this "candidate list" that will
+            advance us in fulfilling the constraints.
+
+        2. Selecting best candidates such that we end up with highest probable candidates
+            that fulfill our constraints.
+        """
```
This should be a comment.

sgugger commented on src/transformers/generation_beam_search.py:
```diff
@@ -330,7 +367,430 @@ def finalize(
+        # need to make new hypothesis that advance the constraints
+        track_new = {"new_seqs": [], "new_states": [], "new_indices": [], "new_tokens": [], "new_scores": []}
+        for seq_idx, pre_seq in enumerate(this_batch_input_ids):
+            """
+            pre_seq = ith sequence generated before this step.
+
+            input_ids -> (topk) generic beam search best model next tokens
+                      -> (advance) constraints forcing the next token
+            either way, we need to sort them into "banks" later, so store a "ConstraintListState" for all types of
+            hypotheses.
+            """
```
This should be a comment.

sgugger commented on src/transformers/generation_beam_search.py:
```diff
@@ -330,7 +367,430 @@ def finalize(
+                        track_new["new_scores"].append(this_batch_token_scores[seq_idx].take(advance_token))
+                        track_new["new_states"].append(new_state)
+            elif push_progress:
+                """
+                Basically, `sent_beam_indices` often chooses very little among `input_ids` the generated sequences that
+                actually fulfill our constraints. For example, let constraints == ["loves pies"] and
+
+                    pre_seq_1 = "The child loves pies and" pre_seq_2 = "The child plays in the playground and"
+
+                Without this step, if `sent_beam_indices` is something like [1,1], then
+                    1. `pre_seq_1` won't be added to the list of (topk) hypothesis since it's not in the indices and
+                    2.  it won't be added to the list of (advance) hypothesis since it's completed already. (this is
+                        the else part of `if constraints_completed[seq_idx]`)
+                    3. it ends up simply getting removed from consideration.
+
+                #3 might be fine and actually desired, since it's likely that it's a low-probability output anyways,
+                especially if it's not in the list of `sent_beam_indices`. But this often leads to lengthened beam
+                search times, since completed sequences keep getting removed after all this effort for constrained
+                generation.
+
+                Here, we basically take `pre_seq_1` and to "push" it into the considered list of hypotheses, by simply
+                appending the next likely token in the vocabulary and adding it to the list of hypotheses.
+                """
```
This should be a comment.

sgugger commented on src/transformers/generation_beam_search.py:
```diff
@@ -330,7 +367,430 @@ def finalize(
+            zipped = all_banks * 100 + all_scores
+            indices = zipped.sort(descending=True).indices
+            sorted_banks = all_banks[indices]
+            """
+            Then we end up with {sorted among bank C}, {sorted among bank C-1}, ..., {sorted among bank 0}
+            """
```
This should be a comment.

sgugger commented on tests/test_generation_beam_search.py:
```diff
@@ -232,6 +233,275 @@ def check_beam_scores_finalize(self, input_ids, next_tokens, next_indices, next_
+        # make sure corresponding score is as good as possible to surely be picked first
+        next_scores[0, 0] = 0.0
+
+        # # because constrainted beam search can't possibly fulfill the constraints in one pass
+        # # especially if constraints involves more than one token, so we repeat this several times.
+        # # doesn't *really* matter that we're not adjusting the scores & tokens TBH.
+        # repeat = 10
+        # for _ in range(repeat):
```
To clean up?

sgugger commented on tests/test_generation_utils.py:
```diff
@@ -1254,6 +1498,24 @@ def _check_encoder_hidden_states_for_generate(self, hidden_states, batch_size, c
+                flag = True
+                break
+
+        assert flag
```
Use `self.assertTrue` here.

sgugger reviewed (approved):
Thanks a lot for your PR! It's great that you put so many comments in your code, but we use multi-line comments inside the code, not triple-quoted strings. Those will be confused with docstrings by our utils, so you should switch them back to comments.

cwkeam committed:
fixing based on comments on PR

cwkeam committed:
took out the testing code that should but work fails without the beam search moditification ; style changes

cwkeam commented:
@patrickvonplaten @sgugger Thanks for the review! I've adjusted all the code accordingly.

As for promoting these changes, I think I'll save it for the `DisjunctiveConstraint` and simple integration to the `generate()` function is complete. I'll open a new PR about it soon. In terms of after those other changes are done, what would be a nice way to promote this change? Could it be possible to write a blog post in huggingface blogs as well as posting on TowardsDataScience about it?

Thanks so much for everyone's help!

patrickvonplaten commented:
> @patrickvonplaten @sgugger Thanks for the review! I've adjusted all the code accordingly.
> 
> As for promoting these changes, I think I'll save it for the `DisjunctiveConstraint` and simple integration to the `generate()` function is complete. I'll open a new PR about it soon. In terms of after those other changes are done, what would be a nice way to promote this change? Could it be possible to write a blog post in huggingface blogs as well as posting on TowardsDataScience about it?
> 
> Thanks so much for everyone's help!

A blog post would be a great idea indeed! This is a really nice feature to write a blog post about I think. Happy to help you

patrickvonplaten reviewed (approved):
Awesome - thanks a lot for reverting this change and opening an issue instead. 
As a final step, could you maybe try to apply @sgugger's comments about docstrings and comments? :-) Happy to help / correct it if you have questions

cwkeam committed:
fixing comments issues

cwkeam committed:
docstrings for ConstraintListState

cwkeam committed:
typo in PhrsalConstraint docstring

cwkeam committed:
Merge branch 'huggingface:master' into constrained_beam_search

cwkeam commented:
@patrickvonplaten 

I've made all the requested adjustments on the comments. 

The test is failing right now at just one script:

```
tests/test_modeling_speech_encoder_decoder.py::Wav2Vec2Speech2Text2::test_encoder_decoder_model_generate
```

which actually runs fine on my machine. Not sure why it's throwing this error after pushing a new commit with adjustments to nothing but comments. This has happened before but simply worked out after fetching from master. How does one generally deal with these problems?

cwkeam committed:
docstrings improvements

patrickvonplaten commented:
Incredible work here @cwkeam - this was one of the cleanest and highest quality PRs that I've seen in a long time! Merging - let's try to tackle disjunctive decoding next :rocket:
